src/runtime_src/core/pcie/driver/linux/xocl/xocl_subdev.c

```diff
@@ -97,43 +100,203 @@ static struct xocl_subdev *xocl_subdev_reserve(xdev_handle_t xdev_hdl,
+	int i, j;
+
+	core = xocl_get_xdev(pldev);
+
+	for (j = 0; j < XOCL_SUBDEV_NUM; j++)
+		for (i = 0; i < XOCL_SUBDEV_MAX_INST; i++)
+			if (core->subdevs[j][i].pldev == pldev)
+				return &core->subdevs[j][i];
+
+	return NULL;
+}
+
+static void xocl_subdev_update_info(xdev_handle_t xdev_hdl,
+	struct xocl_subdev_info *info_array, int *num,
+	struct xocl_subdev_info *sdev_info)
+{
+	int	i, idx;
+
+	for (i = 0; i < *num; i++) {
+		if (info_array[i].id == sdev_info->id &&
+			info_array[i].override_idx == sdev_info->override_idx &&
+			!info_array[i].multi_inst) {
+			memcpy(&info_array[i], sdev_info,
+				sizeof(*info_array));
+			return;
+		}
+
+		if (info_array[i].id > sdev_info->id)
+			break;
+	}
+	idx = i;
+
+	*num += 1;
+	for (i = *num; i > idx; i--) {
+		memcpy(&info_array[i], &info_array[i - 1],
+				sizeof(*info_array));
+	}
+	memcpy(&info_array[idx], sdev_info, sizeof(*info_array));
+}
+
+static struct xocl_subdev_info *xocl_subdev_get_info(xdev_handle_t xdev_hdl,
+		int *num)
+{
+	struct xocl_subdev_info *subdev_info;
+	struct xocl_dev_core *core = xdev_hdl;
+	int i, subdev_num = 0;
+
+	subdev_info = vzalloc(sizeof(*subdev_info) *
+		(core->dyn_subdev_num + core->priv.subdev_num));
+	if (!subdev_info)
+		return NULL;
+
+	/* construct subdevice info array */
+	for (i = 0; i < core->priv.subdev_num; i++)
+		xocl_subdev_update_info(core, subdev_info, &subdev_num,
+				&core->priv.subdev_info[i]);
+
+	for (i = 0; i < core->dyn_subdev_num; i++) {
+		if (core->dyn_subdev_store[i].pf != XOCL_PCI_FUNC(core))
+			continue;
+		xocl_subdev_update_info(core, subdev_info, &subdev_num,
+				&core->dyn_subdev_store[i].info);
+	}
+	if (subdev_num <= 0) {
+		vfree(subdev_info);
+		*num = 0;
+		return NULL;
+	}
+
+	*num = subdev_num;
+
+	return subdev_info;
+
+}
+
+static dev_t xocl_subdev_get_devt(struct platform_device *pldev)
+{
+	struct xocl_dev_core	*core;
+	struct xocl_subdev *subdev;
+
+	core = xocl_get_xdev(pldev);
+
+	subdev = xocl_subdev_lookup(pldev);
+	if (!subdev) {
+		xocl_err(&pldev->dev, "did not find subdev");
+		return -1;
+	}
+
+	return MKDEV(MAJOR(XOCL_GET_DRV_PRI(pldev)->dev),
+		(subdev->inst & MINORMASK));
+}
+
+static int xocl_subdev_cdev_create(struct platform_device *pdev,
+		struct cdev **cdev)
+{
+	struct xocl_dev_core *core;
+	struct device *sysdev;
+	struct cdev *cdevp;
+	int ret;
+
+	if (!XOCL_GET_DRV_PRI(pdev) || !XOCL_GET_DRV_PRI(pdev)->fops)
+		return 0;
+
+	if (!platform_get_drvdata(pdev)) {
+		xocl_err(&pdev->dev, "driver did not probe");
+		return -EAGAIN;
+	}
+
+	core = xocl_get_xdev(pdev);
+	cdevp = cdev_alloc();
+	if (!cdevp) {
+		xocl_err(&pdev->dev, "alloc cdev failed");
+		ret = -EFAULT;
+		goto failed;
+	}
+
+	cdevp->ops = XOCL_GET_DRV_PRI(pdev)->fops;
+	cdevp->owner = THIS_MODULE;
+	cdevp->dev = xocl_subdev_get_devt(pdev);
+
+	ret = cdev_add(cdevp, cdevp->dev, 1);
+	if (ret) {
+		xocl_err(&pdev->dev, "cdev add failed %d", ret);
+		goto failed;
+	}
+
+	if (XOCL_GET_DRV_PRI(pdev)->cdev_name)
```

Comment:
i think we should also create mailbox and nifd dev node using new schema. they are completely under our control after all and can be easily changed. as for xvc, can we create both dev nodes (old and new schema) so that user could migrate?